html2asketch/nodeToSketchLayers.js

```diff
@@ -247,12 +248,27 @@ export default async function nodeToSketchLayers(node) {
         fixY = (textBCR.height - lineHeightInt * numberOfLines) / 2;
       }
 
+      let textValue = textNode.nodeValue;
+
+      switch (whiteSpace) {
+        case 'normal':
+        case 'nowrap':
+          textValue = textValue.trim().replace(/\n/g, ' ').replace(/ +/g, ' ');
```

Comment:
maybe instead of using these `.replace` regexes, we should use innerText? e.g.

```
  <p> adam     west is
    the best
    
    
    !!!
    
  </p>
```

<img width="215" alt="screen shot 2017-12-02 at 19 52 27" src="https://user-images.githubusercontent.com/985504/33518722-595d0f66-d79a-11e7-84da-5e9b359e1b2b.png">

WDYT?

Reply:
`.innerText` is nice because it respects `white-space` (and probably other CSS properties that can influence how text is rendered):

<img width="172" alt="screen shot 2017-12-02 at 19 59 00" src="https://user-images.githubusercontent.com/985504/33518776-49b09406-d79b-11e7-90da-dc30a3747787.png">
<img width="202" alt="screen shot 2017-12-02 at 19 58 44" src="https://user-images.githubusercontent.com/985504/33518780-500eb0d0-d79b-11e7-85a3-ebb38e6b18f7.png">

---

<img width="178" alt="screen shot 2017-12-02 at 19 58 51" src="https://user-images.githubusercontent.com/985504/33518782-5723d7ec-d79b-11e7-8d58-2828921fda9f.png">

<img width="170" alt="screen shot 2017-12-02 at 19 58 41" src="https://user-images.githubusercontent.com/985504/33518783-5cca8d8a-d79b-11e7-8c57-670669f6f4ba.png">

---

The only issue is that `.innerText` doesn't work on text nodes, it works on elements and elements can have multiple children e.g. `<p>some text <span> other text</span> <img /></p>`. So, I'm not sure if we can use it 🤔

Reply:
If we give up `textNode` for `innerText`, I think we should solve more problems.
So, I want to keep the current way.

If the problem of width is solved, Is this PR a not bad solution?.. 🙄

---

And I tried experiments for the difference of width.
The differences were not related to style attributes such as width, font-size, letter-spacing, and font-family ....
It just always needed an additional width of 0.01 ~ 0.1.
I have not found a difference greater than 0.1 yet.

So I suggest just adding a value of 0.1

```javascript
const text = new Text({
  x: textBCR.x,
  y: textBCR.y + fixY,
  width: textBCR.width + 0.1,
  height: textBCR.height,
  text: textValue,
  style: textStyle,
  multiline: numberOfLines > 1
});
```

What do you think?

---

And I have a question.
When measuring the `clientRect` of a `text node`, why use `calculateBCRFromRanges` instead of` rangeHelper.getBoundingClientRect() `?

I could not find a problem when I replaced it with `rangeHelper.getBoundingClientRect()`.  

```javascript
const textBCR = rangeHelper.getBoundingClientRect();
```

Reply:
> If we give up textNode because of innerText, I think we should solve more problems.

So, I want to keep the current way. ㅠㅠ

Not sure if I understand. In your opinion we should stick to `textContent`?

> So I suggest just adding a value of 0.1

It does sound like a hack, but if we don't know what exactly is causing this, we will have to do it as you suggest 👍 Lets have a separate PR for this change though.

> Why use `calculateBCRFromRanges` instead of `rangeHelper.getBoundingClientRect()`?

TBH I don't remember. Can you create a separate PR for this so I can play with it?

Reply:
Yes, my opinion is to use 'textNode.nodeValue'. like now.

I will write a new PR for two other things.

Reply:
:+1:

One more question, your regexes only take space character into account. How about tab, carrige return and other whitespace characters? I think that they all behave the same in HTML. Maybe instead of ` ` we should use `\s`?

Reply:
Yes,.. The current code has problem because of `tabs` and other `whitespace`.

But, `\s` contains` \n`.
RegExp in this code removes whitespace or makes one. It should not contain `\n`.

This is the code I can think. This does not look clean...
Is this okay?....

```javascript
textValue = textValue.trim().replace(/\n/g, ' ').replace(/[ \f\r\t\v\u00a0\u1680\u180e\u2000-\u200a\u2028\u2029\u202f\u205f\u3000]+/g, ' ');

```

I don't know how to make RegExp efficiently ...ㅠ sorry..

Reply:
I found this trick with double negative:`[^\S\n]` - which means "not non-whitespace character and not \n"

Reply:
cool! It works well. 😁